modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportPutDatabaseConfigurationAction.java

```diff
@@ -0,0 +1,198 @@
+import org.elasticsearch.ingest.geoip.direct.PutDatabaseConfigurationAction.Request;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class TransportPutDatabaseConfigurationAction extends TransportMasterNodeAction<Request, AcknowledgedResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportPutDatabaseConfigurationAction.class);
+
+    private static final SimpleBatchedExecutor<UpdateDatabaseConfigurationTask, Void> UPDATE_TASK_EXECUTOR = new SimpleBatchedExecutor<>() {
+        @Override
+        public Tuple<ClusterState, Void> executeTask(UpdateDatabaseConfigurationTask task, ClusterState clusterState) throws Exception {
+            return Tuple.tuple(task.execute(clusterState), null);
+        }
+
+        @Override
+        public void taskSucceeded(UpdateDatabaseConfigurationTask task, Void unused) {
+            logger.trace("Updated cluster state for creation-or-update of database configuration [{}]", task.database.id());
+            task.listener.onResponse(AcknowledgedResponse.TRUE);
+        }
+    };
+
+    private final MasterServiceTaskQueue<UpdateDatabaseConfigurationTask> updateDatabaseConfigurationTaskQueue;
+
+    @Inject
+    public TransportPutDatabaseConfigurationAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            PutDatabaseConfigurationAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            Request::new,
+            indexNameExpressionResolver,
+            AcknowledgedResponse::readFrom,
+            EsExecutors.DIRECT_EXECUTOR_SERVICE
+        );
+        this.updateDatabaseConfigurationTaskQueue = clusterService.createTaskQueue(
+            "update-geoip-database-configuration-state-update",
+            Priority.NORMAL,
+            UPDATE_TASK_EXECUTOR
+        );
+    }
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener) {
+        final String id = request.getDatabase().id();
+
+        {
+            IngestGeoIpMetadata geoIpMeta = state.metadata().custom(IngestGeoIpMetadata.TYPE, IngestGeoIpMetadata.EMPTY);
+
+            final DatabaseConfigurationMetadata existingDatabase = geoIpMeta.getDatabases().get(id);
+            // make the request a no-op if the databases match exactly
+            if (isNoopUpdate(existingDatabase, request.getDatabase())) {
+                listener.onResponse(AcknowledgedResponse.TRUE);
+                return;
+            }
```

Comment:
I think bc7e4eb4e75 addresses this, but your concern is intricate enough that I'm leaving the resolve unresolved so that you can choose to resolve if you're happy with the new logic.